#14 - UI Functionality and Integrating Damage Logic

overseer25 opened the pull request:
## Overview

The purpose of this work is to add basic functionality to the following UI elements:

- Heads-up Display
- Pause Menu

Furthermore, as the player health bar required working damage logic to test, I integrated the damage models into the player, the player weapons, and enemies.

### Heads-up Display

Created a script for controlling the HUD. The heads-up display now syncs up with the player's health and ammo. Different ammo types have different ammo icons, as certain weapons may share ammo types. The player's health bar will change color depending on how much health the player has remaining.

### Pause Menu

The pause menu now works. When the player hits the "pause" key, the game will pause and display the pause menu. Hitting the resume button or clicking the "pause key" again will continue the game. The In-game console can appear regardless of whether the game is paused or not, so logic was added to make sure the console and pause menu interact well. When playing a build, the Quit button will exit the game. The GameManager was given the functionality to quit the game, so all quit sources utilize this logic.

### Damage

The player's weapons now contain a DamageInfo field. This damage info is what gets sent to entities that can take damage. The HordeBot also contains a DamageInfo field, and uses this data when dealing damage to the player.

### Health Controllers

I reworked the health controller so that there exists a base class that all health controllers can utilize. Created a separate player health controller and enemy health controller.

packerbacker8 commented on Assets/Scripts/Controllers/BaseHealthController.cs:
```diff
@@ -0,0 +1,139 @@
+
+    public bool IsDead() 
+    {
+        return currentHealth == minimumHealth;
```
I know we want to clamp current health to always be between max and min health. But in case we do something dumb with our clamping on accident, do we want to have this check be `currentHealth <= minimumHealth` just to make sure we catch if they are dead?

overseer25 replied:
If we aren't confident in our abilities to clamp the health, then clamping should probably just be removed altogether. Otherwise, we should assume clamping is working as intended.

packerbacker8 commented on Assets/Scripts/Player/PlayerHealthController.cs:
```diff
@@ -0,0 +1,53 @@
+    /// </summary>
+    /// <param name="healthAdjustmentType"></param>
+    /// <param name="amount"></param>
+    public override void AdjustCurrentHealth(HEALTH_ADJUSTMENT healthAdjustmentType, int amount)
+    {
+        previousHealth = currentHealth;
+        amount = Mathf.Abs(amount);
+
+        if(healthAdjustmentType == HEALTH_ADJUSTMENT.DAMAGE) {
+            if (isInvulnerable) return;
+
+            currentHealth = Mathf.Clamp(currentHealth - amount, minimumHealth, maximumHealth);
+            if (IsDead())
+            {
+                Kill();
+                return;
+            }
+        }
+        else if (healthAdjustmentType == HEALTH_ADJUSTMENT.HEAL){
+            currentHealth = Mathf.Clamp(currentHealth + amount, minimumHealth, maximumHealth);
+        }
+        hud.UpdateHealthBar(currentHealth);
+    }
```
This function seems to be pretty similar between player and AI. I am wondering if maybe we want it to be an virtual method that they can override but have the base virtual method do some of that generic health stuff between them? if they end up being vastly different we can always just have the override take care of it like it is now

overseer25 commented:
Addressed PR comments. @packerbacker8 is there a reason you decided to leave comments for some suggestions, but actually push changes for others? I would urge against pushing commits to a branch that is not yours, unless you are a co-owner of that branch.

packerbacker8 commented:
I dont know if I had a "good" reason, but I remembered it was a thing and wanted to try it out. Figured it worked and no reason to have someone else do the work if it was done with what I was trying
